app/scripts/controllers/network/fake-provider.ts

```diff
@@ -0,0 +1,224 @@
+ * @property delay - The amount of time that will pass after the callback is
+ * called with the response.
+ * @property discardAfterMatching - Usually after the stub matches a request, it
+ * is discarded, but setting this to true prevents that from happening. True by
+ * default.
+ */
+export type FakeProviderStub<Params, Response> = {
+  request: {
+    method: string;
+    params?: Params[];
+  };
+  delay?: number;
+  discardAfterMatching?: boolean;
+  beforeCompleting?: () => void | Promise<void>;
+} & (
+  | {
+      response:
+        | (Omit<JsonRpcSuccess<Response>, 'id' | 'jsonrpc' | 'result'> & {
+            result: Response;
+          })
+        | (Omit<JsonRpcFailure, 'id' | 'jsonrpc'> & { error: string });
+    }
+  | {
+      error: unknown;
+    }
+  | {
+      implementation: () => void;
+    }
+);
+
+/**
+ * The set of options that the FakeProviderEngine constructor takes.
+ *
+ * @property stubs - A set of objects that allow specifying the behavior
+ * of specific invocations of `sendAsync` matching a `method`.
+ */
+interface FakeProviderEngineOptions {
+  stubs?: FakeProviderStub<any, any>[];
+}
+
+/**
+ * FakeProviderEngine is an implementation of the provider that
+ * NetworkController exposes, which is actually an instance of
+ * Web3ProviderEngine (from the `web3-provider-engine` package). Hence it
+ * supports the same interface as Web3ProviderEngine, except that fake responses
+ * for any RPC methods that are accessed can be supplied via an API that is more
+ * succinct than using Jest's mocking API.
+ */
+// NOTE: We shouldn't need to extend from the "real" provider here, but
+// we'd need a `SafeEventEmitterProvider` _interface_ and that doesn't exist (at
+// least not yet).
+export class FakeProvider extends SafeEventEmitterProvider {
+  calledStubs: FakeProviderStub<any, any>[];
+
+  #originalStubs: FakeProviderStub<any, any>[];
+
+  #stubs: FakeProviderStub<any, any>[];
+
+  /**
+   * Makes a new instance of the fake provider.
+   *
+   * @param options - The options.
+   * @param options.stubs - A set of objects that allow specifying the behavior
+   * of specific invocations of `sendAsync` matching a `method`.
+   */
+  constructor({ stubs = [] }: FakeProviderEngineOptions) {
+    super({ engine: new JsonRpcEngine() });
+    this.#originalStubs = stubs;
+    this.#stubs = this.#originalStubs.slice();
+    this.calledStubs = [];
+  }
+
+  send = <Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  sendAsync = <Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  #handleSend<Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback?: (error: unknown, response?: JsonRpcResponse<Response>) => void,
```

Comment:
It seems that a callback is passed in both places where this is called. Should this be mandatory rather than optional?

Reply:
Yes, this is a bit funky. I also realized that the signature of `sendAsync` was incorrect. I've updated it and aligned the other methods here: [424320f](https://github.com/MetaMask/metamask-extension/pull/18628/commits/424320f7b3dd3de60b1954d0b3643d6c91ef92ab)

Reply:
Hmm, how was it incorrect? The old one looks like it matches the safe event emitter provider: https://github.com/MetaMask/eth-json-rpc-provider/blob/3fa7c4f2cdcf9ea0c77b77c8e71c9cf6194ba2b0/src/safe-event-emitter-provider.ts#L38

That is, the callback is always required, not optional.

Reply:
Yeah... okay. I saw that `send` throws if it's not given a callback but `sendAsync` does not. It's inconsistent, and maybe there is a good reason for it, but I guess that doesn't mean that the callback for `sendAsync` is optional. I've made them all required: [08f5d7d](https://github.com/MetaMask/metamask-extension/pull/18628/commits/08f5d7dfc3ebbfc2b3e8a80c4964bca473e5172b)